docs/source/deprecation.md

```diff
@@ -1,14 +1,164 @@
-## Deprecation
+# Deprecation
+
+The deprecation process is here to make (your) life of developers easier.
+It's not possible to write perfect solution for everything and as the project
+is evolving, it happens that some functionality needs to be changed, replaced
+or dropped completely. Such situations are inevitable. To reduce negative
+impact on your code, we introduce the deprecation process described below.
+
+## What does it mean that something is deprecated?
+
+When you deprecate something, the only thing that changes is that the
+deprecated entity is marked in the code as deprecated which can have
+additional impact,likemessages produced on the user's terminal,
+in the report, ... But the rest of the functionality is the same as before,
+until the entity is removed completely.
+
+## What is the deprecation process for leapp
+
+In case some public public functions or models are to be removed for any
+reason, these need to be marked as deprecated before the removal (if possible).
+In our projects, the deprecation will be applied only for functions and models
+that have been introduced in an official release (IOW, a functionality that
+has been merged into the upstream, but has been removed before the release
+or was marked as experimental all the time, is going to be removed without
+the deprecation state). The time period during which the deprecated entity
+won't be removed is at least 6 months. That doesn't mean we will remove
+everything deprecated immadiately after the 6 months, but it's to be expected
+that it will be dropped anytime between 6 and 12 months since the deprecation.
+
+In case of issues, deprecated functionality is not going to be fixed since
+it is deprecated (unless it is fixed e.g. as a side-effect of another
+problem fix).
+
+## How does leapp informs about deprecated functionality?
+
+Mainly via release notes and changelogs. In the official leapp related projects
```

Comment:
Why not to be more specific here: Via release notes, changelogs, deprecation warnings (included to the upgrade report and pytest captured warning summary)

Reply:
Because here we present what is the real source of truth. We cannot expect that people will run pytest for their custom code. We cannot expect that the report message will be generated everytime. It's described in limitations of our solution below.